### ISSUE TITLE
Figure 6 from your paper.

### ISSUE THREAD

**Issue by Jiafei127** [NeonLeexiang/DLGSANet#12]
The ReLU function replacing Softmax is of strong interest to me in terms of the sparsity it can provide in that global attention mechanism. Could the authors please provide the visualisation code of Figure 6 in your paper for our learning?

**Comment by Jiafei127** [comment]
跪求大佬指导，感激涕零！！ @NeonLeexiang

**Comment by NeonLeexiang** [comment]
非常抱歉这么迟才回复您，

关于 Figure 6 特征图的获取是借用了 `pytorch` 的 Hook 机制构建`HookTool` Function。得益于BasicSR标准的框架设计，分离的Model和architecture，所以可以通过在model中构建`HookTool()`来实现 get feature by hook 操作。

比如：

```python
class HookTool: 
    def __init__(self):
        self.fea = None 

    def hook_fun(self, module, fea_in, fea_out):
        self.fea = fea_out

# register
def get_feas_by_hook(model):
    fea_hooks = []
    
    # you can use modules' name to select the specific feature layers
    for n, m in model.named_modules():		 
        if isinstance(m, torch.nn.Conv2d):	# 
            cur_hook = HookTool()
            m.register_forward_hook(cur_hook.hook_fun)
            fea_hooks.append(cur_hook)
    return fea_hooks

```

关于 `pytorch hook` 机制的使用可以google it～ 我也是当时现学现卖的。

获取到对应的layers的feature之后，可以借助`torchvision.util ` 中的 `make_grid()` 结合  `tensorboardX` 将其可视化出来，或者使用其他的方式进行可视化都可，比较麻烦的可能就是 `feature hook` 这个步骤。

_关于 `torchvision.util.make_grid()` 的内容可借助官方文档 [`[torchvision](https://pytorch.org/vision/stable/generated/torchvision.utils.make_grid.html)`]( https://pytorch.org/vision/stable/generated/torchvision.utils.make_grid.html)_


---

Sorry about the delayed response.

Regarding Figure 6, PyTorch's Hook mechanism serves as the foundation for the feature visualization. We can incorporate our hook function into the model file to acquire the particular feature layers thanks to BasicSR's standard design.

For example:

```python
class HookTool: 
    def __init__(self):
        self.fea = None 

    def hook_fun(self, module, fea_in, fea_out):
        self.fea = fea_out

# register
def get_feas_by_hook(model):
    fea_hooks = []
    
    # you can use modules' name to select the specific feature layers
    for n, m in model.named_modules():		 
        if isinstance(m, torch.nn.Conv2d):	# 
            cur_hook = HookTool()
            m.register_forward_hook(cur_hook.hook_fun)
            fea_hooks.append(cur_hook)
    return fea_hooks
```

aOnce the features are purchased, we can view the feature using  `torchvision.util.make_grid()`  and `tensorboardX`.

_You may get more information about  `torchvision.util.make_grid()` in  [`[torchvision](https://pytorch.org/vision/stable/generated/torchvision.utils.make_grid.html)`]( https://pytorch.org/vision/stable/generated/torchvision.utils.make_grid.html)_
